python/pyspark/util.py

```diff
@@ -323,6 +323,18 @@ def __del__(self):
                         thread_connection.close()
 
 
+class PySparkWarning(UserWarning):
+    """A base class for PySpark warnings."""
+
+
+class PySparkFutureWarning(PySparkWarning, FutureWarning):
+    """A base class for deprecated PySpark APIs."""
+
+
+class PySparkResourceWarning(PySparkWarning, ResourceWarning):
+    """A base class for PySpark warnings related to resource usage."""
```

Comment:
Hm, what was on my mind was to use directly the built-in warning types like `FutureWarning`or `ResourceWarning`. Would you mind explaining this exception hierarchy and what we can gain from it?

Reply:
Thanks for looking at this @HyukjinKwon.

The type hierarchy I had in mind was something like this, where PySpark-specific warnings inherit from top-level `PySparkWarning` and, optionally, from base Python warning classes.

![class](https://user-images.githubusercontent.com/1554276/103542204-e47a0a00-4e9c-11eb-9b32-db604f2db40c.png)

The reason I thought about this design is to have better control over warning filtering ‒ built-in mechanism, IMHO, is a bit limited here.

However, I am not that convinced that this is necessary and I am happy to limit the scope of this PR to built-in exceptions only (I recognize it might be an over-engineering and the problem I see with warnings is too localized).

Reply:
I see. I think it's okay not to have the base classes for warnings for now - I got that it depends on how individuals think though.

If we need some specific warning types, we could define it and directly inherit from built-in warning classes (and then expose it as a module or pacakge  e.g. [pandas.errors](https://github.com/pandas-dev/pandas/blob/master/pandas/errors/__init__.py)). Or we could even think about introducing the base classes later when we have enough instances to generalize.